### PR TITLE
Memory leak fixes.

### PR DESCRIPTION
**Pull request by bowb**
Fix invalid read on from.
Fix memory leaks when using subject mapper.

There are other issues that need addressed. There are memory leaks and other invalid code that needs fixed.

Fixes https://github.com/OpenSC/pam_pkcs11/issues/65

### CODE REVIEW

File: src/common/pkcs11_lib.c [hunk 1]
```diff
@@ -1585,6 +1585,7 @@ cert_object_t **get_certificate_list(pkcs11_handle_t *h, int *ncerts)
       cert_template[3].pValue = NULL;
       rv = h->fl->C_GetAttributeValue(h->session, object, cert_template, 4);
       if (rv != CKR_OK) {
+        free(id_value);
```
Comment (wolneykien) [inline src/common/pkcs11_lib.c:1588]: Thanks for one more fixed leak! However, wouldn't it be better to free the allocated resources at the _single, common_ exit point? Maybe we should rewrite the function in the style:

```
/* Initialization block */
id_value = NULL;
cert_value = NULL;
...

/* The body */
...
if ... {
    goto getlist_error;
}
if ... {
    goto getlist_error;
}
...

/* Normal exit: */
ret = h->certs;
exit:
    free(id_value);
    free(cert_value);
    return ret;

/* Exit on error: */
getlist_error:
    ret = NULL;
    goto exit;
```

File: src/mappers/subject_mapper.c [hunk 2]
```diff
@@ -62,11 +62,15 @@ static char * subject_mapper_find_user(X509 *x509, void *context, int *match) {
 		DBG("X509_get_subject_name failed");
 		return NULL;
 	}
-	return mapfile_find(filename,entries[0],ignorecase,match);
+	char* val = mapfile_find(filename,entries[0],ignorecase,match);
+	if(entries[0]) {
```
Comment (wolneykien) [inline src/mappers/subject_mapper.c:66]: Shouldn't we have a universal function to free `cert_info()` entries? Something like:

```
while (entries[i]) {
    free(entries[i]);
}
```

Freeing just the 0-th element looks strange… What if there would be more than one element?

Comment (bowb) [reply src/mappers/subject_mapper.c:66]: There might be a bigger issue with this one. The pattern that is used in `cert_info.c` doesn't track the possible size of the `entries` allocated and there might be other memory leaks going on with other functions in the file. I just tracked memory leaks for `subject_mapping` and ended up in the `cert_info_subject` function and took note that it only allocated a single entry. I think a re-factor will need to be done to return the actual count of entries allocated. I thought it was strange that `cert_info_subject` is using

File: src/common/pkcs11_lib.c (hunk 1 above)
Comment (bowb) [reply src/common/pkcs11_lib.c]: Updated

**Comment by wolneykien** [comment]
Hi! Please, look at https://github.com/bowb/pam_pkcs11/pull/1 .

**Comment by bowb** [comment]
Bump

**Review by wolneykien — APPROVED** [review]

**Comment by wolneykien** [comment]
Sorry for keeping you waiting...

**Pull request merged**